libminifi/src/utils/NetworkInterfaceInfo.cpp

```diff
@@ -0,0 +1,169 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "utils/NetworkInterfaceInfo.h"
+#include "utils/OsUtils.h"
+#include "core/logging/LoggerConfiguration.h"
+#ifdef WIN32
+#include <Windows.h>
+#include <winsock2.h>
+#include <iphlpapi.h>
+#include <WS2tcpip.h>
+#pragma comment(lib, "IPHLPAPI.lib")
+#else
+#include <unistd.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <net/if.h>
+#include <arpa/inet.h>
+#include <ifaddrs.h>
+#endif
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+std::shared_ptr<core::logging::Logger> NetworkInterfaceInfo::logger_ = core::logging::LoggerFactory<NetworkInterfaceInfo>::getLogger();
+
+#ifdef WIN32
+namespace {
+std::string utf8_encode(const std::wstring& wstr) {
+  if (wstr.empty())
+    return std::string();
+  int size_needed = WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), -1, nullptr, 0, nullptr, nullptr);
+  std::string result_string(size_needed, 0);
+  WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), -1, result_string.data(), size_needed, nullptr, nullptr);
+  return result_string;
+}
+}
+
+NetworkInterfaceInfo::NetworkInterfaceInfo(const IP_ADAPTER_ADDRESSES* adapter) {
+  name_ = utf8_encode(adapter->FriendlyName);
+  for (auto unicast_address = adapter->FirstUnicastAddress; unicast_address != nullptr; unicast_address = unicast_address->Next) {
+    if (unicast_address->Address.lpSockaddr->sa_family == AF_INET) {
+      ip_v4_addresses_.push_back(OsUtils::sockaddr_ntop(unicast_address->Address.lpSockaddr));
+    } else if (unicast_address->Address.lpSockaddr->sa_family == AF_INET6) {
+      ip_v6_addresses_.push_back(OsUtils::sockaddr_ntop(unicast_address->Address.lpSockaddr));
+    }
+  }
+  running_ = adapter->OperStatus == IfOperStatusUp;
+  loopback_ = adapter->IfType == IF_TYPE_SOFTWARE_LOOPBACK;
+}
+#else
+NetworkInterfaceInfo::NetworkInterfaceInfo(const struct ifaddrs* ifa) {
+  name_ = ifa->ifa_name;
+  if (ifa->ifa_addr->sa_family == AF_INET) {
+    ip_v4_addresses_.push_back(OsUtils::sockaddr_ntop(ifa->ifa_addr));
+  } else if (ifa->ifa_addr->sa_family == AF_INET6) {
+    ip_v6_addresses_.push_back(OsUtils::sockaddr_ntop(ifa->ifa_addr));
+  }
+  running_ = (ifa->ifa_flags & IFF_RUNNING);
+  loopback_ = (ifa->ifa_flags & IFF_LOOPBACK);
+}
```

Comment:
What about `ifa->ifa_next`? This is also a linked list, similarly to the windows version.

Reply:
That part is addressed in getNetworkInterfaceInfos
In linux the ip addresses are in a simple linked list. e.g. adapter1.ipv4->adapter2.ipv4->adapter1.ipv6->adapter2.ipv6
In Windows its a linked list based on the adapters so adapter1(with nested linked list ipv4->ipv6) -> adapter2(with nested linked list ipv4->ipv6)
This is the reason that the getNetworkInterfaceInfos traverses these lists (the windows one populates all of its ipaddresses), while the linux one just creates an adapterinfo with one ipaddress then merges them later.